presto-main/src/main/java/com/facebook/presto/server/security/AuthenticationFilter.java

```diff
@@ -95,9 +107,37 @@ public void doFilter(ServletRequest servletRequest, ServletResponse servletRespo
                 e.getAuthenticateHeader().ifPresent(authenticateHeaders::add);
                 continue;
             }
-
             // authentication succeeded
-            nextFilter.doFilter(withPrincipal(request, principal), response);
+            CustomHttpServletRequestWrapper wrappedRequest = new CustomHttpServletRequestWrapper(request);
+            Map<String, String> extraHeadersMap = new HashMap<>();
+            Set<String> globallyAddedHeaders = new HashSet<>();
+
+            for (ClientRequestFilter requestFilter : clientRequestFilterManager.getClientRequestFilters()) {
+                boolean headersPresent = requestFilter.getHeaderNames().stream()
+                        .allMatch(headerName -> request.getHeader(headerName) != null);
+
+                if (!headersPresent) {
+                    Optional<Map<String, String>> extraHeaderValueMap = requestFilter.getExtraHeaders(principal);
+
+                    extraHeaderValueMap.ifPresent(map -> {
+                        for (Map.Entry<String, String> extraHeaderEntry : map.entrySet()) {
+                            String headerKey = extraHeaderEntry.getKey();
+                            if (headersBlockList.contains(headerKey)) {
+                                throw new RuntimeException("Modification attempt detected: The header " + headerKey + " is present in the blocked headers list.");
```

Comment:
Please throw a PrestoException and introduce a new error code for this.  It should be a system error.